distributed_shampoo/shampoo_fsdp_utils.py

```diff
@@ -0,0 +1,900 @@
+LICENSE file in the root directory of this source tree.
+
+"""
+
+import enum
+import logging
+from copy import deepcopy
+from math import prod
+from typing import List, Tuple, Union
+
+import numpy as np
+import torch
+import torch.distributed as dist
+from torch import Tensor
+
+from distributed_shampoo.shampoo_utils import (
+    AdagradPreconditioner,
+    BlockShampooPreconditioner,
+    DistributedPreconditioner,
+    GraftingType,
+    LargeDimMethod,
+    ShampooPreconditioner,
+)
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: think about enum names
+
+
+class ConvexShapeRecoveryMethod(enum.Enum):
+    SPLIT = 0
+    COMM = 1
+
+
+class CommunicationType(enum.Enum):
+    SEND = 0
+    RECV = 1
+    NONE = 2
+
+
+class CommunicationDirection(enum.Enum):
+    FORWARD = 0
+    BACKWARD = 1
+
+
+def convex_split(
+    tensor: Tensor,
+    orig_shape: torch.Size,
+    start_idx: int,
+    end_idx: int,
+    include_placeholder: bool = False,
+) -> List[Tensor]:
+    """Chunks tensor across dimensions row-wise to be convex.
+
+    Starting from the leftmost dimension, the largest possible slices in each dimension
+    (with the remaining dimensions on the right retaining the original shape) are split off.
+
+    2D example:
+     _______________                  _______________
+    |       ________|                |       ________|
+    |______|        |                |______|________|
+    |               |                |               |
+    |    shard  ____|       ->       |_______________|
+    |__________|    |                |__________|    |
+    |               |                |               |
+    |_______________|                |_______________|
+
+    Args:
+        tensor (Tensor): Flattened gradient or tensor to split.
+        orig_shape (torch.Size): Shape of original tensor that tensor is a slice of.
+        start_idx (int): Flattened index in original tensor where tensor starts.
+        end_idx (int): Flattened index in original tensor where tensor ends (inclusive).
+        include_placeholder (bool): Whether to include empty partitioned tensors as placeholders. (Default: False)
+
+    Returns:
+        split_tensors (List[Tensor]): List of tensors.
+
+    """
+    if len(tensor.size()) != 1:
+        logger.info(
+            f"Input tensor is not flat, has shape {tensor.size()}. Continuing without splitting."
+        )
+        return tensor
+
+    end_idx += 1  # correct off-by-one (FSDP shard_param_info provides inclusive index)
+    assert (
+        end_idx - start_idx == tensor.size()[0]
+    ), f"Start/end indices do not match tensor size: start {start_idx} end {end_idx}, tensor size {tensor.size()}"
+
+    # maintains the order that partitions had in the flat tensor
+    split_tensors_left = []
+    split_tensors_right = []
+    left_idx = None
+    right_idx = None
+    center_partition = False
+    for i in range(1, len(orig_shape) + 1):
```

Comment:
Nit: Let's add a space here? Easier for the eyes. :P
```suggestion
    center_partition = False
    
    for i in range(1, len(orig_shape) + 1):
```